soroban-env-host/src/vm.rs

```diff
@@ -220,87 +220,98 @@ impl Vm {
+            };
+
+            Self::check_max_args(host, &module)?;
+            Self::check_meta_section(host, &module)?;
+
+            let mut store = Store::new(&engine, host.clone());
+            store.limiter(|host| host);
+
+            let mut linker = <Linker<Host>>::new(&engine);
 
-        let mut store = Store::new(&engine, host.clone());
-        store.limiter(|host| host);
+            {
+                let _span0 = tracy_span!("define host functions");
+                for hf in HOST_FUNCTIONS {
+                    let func = (hf.wrap)(&mut store);
+                    host.map_err(
+                        linker
+                            .define(hf.mod_str, hf.fn_str, func)
+                            .map_err(|le| wasmi::Error::Linker(le)),
+                    )?;
+                }
+            }
 
-        let mut linker = <Linker<Host>>::new(&engine);
+            let not_started_instance = {
+                let _span0 = tracy_span!("instantiate module");
+                host.map_err(linker.instantiate(&mut store, &module))?
+            };
 
-        {
-            let _span0 = tracy_span!("define host functions");
-            for hf in HOST_FUNCTIONS {
-                let func = (hf.wrap)(&mut store);
-                host.map_err(
-                    linker
-                        .define(hf.mod_str, hf.fn_str, func)
-                        .map_err(|le| wasmi::Error::Linker(le)),
-                )?;
-            }
-        }
+            let instance = host.map_err(
+                not_started_instance
+                    .ensure_no_start(&mut store)
+                    .map_err(|ie| wasmi::Error::Instantiation(ie)),
+            )?;
 
-        let not_started_instance = {
-            let _span0 = tracy_span!("instantiate module");
-            host.map_err(linker.instantiate(&mut store, &module))?
+            let memory = if let Some(ext) = instance.get_export(&mut store, "memory") {
+                ext.into_memory()
+            } else {
+                None
+            };
+
+            // Here we do _not_ supply the store with any fuel. Fuel is supplied
+            // right before the VM is being run, i.e., before crossing the host->VM
+            // boundary.
+            Ok(Rc::new(Self {
+                contract_id,
+                module,
+                store: RefCell::new(store),
+                instance,
+                memory,
+            }))
         };
 
-        let instance = host.map_err(
-            not_started_instance
-                .ensure_no_start(&mut store)
-                .map_err(|ie| wasmi::Error::Instantiation(ie)),
-        )?;
+        if cfg!(feature = "time") {
+            let now = Instant::now();
+            let vm = instantiate()?;
 
-        let memory = if let Some(ext) = instance.get_export(&mut store, "memory") {
-            ext.into_memory()
+            host.as_budget().track_time(
```

Comment:
Probably it would make sense to guard `track_time` with the same feature (it's meant to only be used in conjunction with real time).